src/core/execution/TradeShipExecution.ts

```diff
@@ -94,25 +91,46 @@ export class TradeShipExecution implements Execution {
       }
     }
 
-    const result = this.pathFinder.nextTile(
-      this.tradeShip.tile(),
-      this._dstPort.tile(),
-    );
+    const currentTile = this.tradeShip.tile();
+    const dstPort = this._dstPort;
+
+    
+    if (dstPort.data.pathCache) {
+      const key = `${this.mg.x(currentTile)},${this.mg.y(currentTile)}`;
+      const cachedNextTile = dstPort.data.pathCache.get(key);
+      if (cachedNextTile) {
+        
+        this.tradeShip.move(cachedNextTile);
+        if (this.mg.isWater(cachedNextTile) && this.mg.isShoreline(cachedNextTile)) {
+          this.tradeShip.setSafeFromPirates();
+        }
+        return;
+      }
+    }
+
+    
+    const result = this.pathFinder.nextTile(currentTile, dstPort.tile());
 
     switch (result.type) {
       case PathFindResultType.Completed:
         this.complete();
         break;
       case PathFindResultType.Pending:
-        // Fire unit event to rerender.
-        this.tradeShip.move(this.tradeShip.tile());
+        this.tradeShip.move(this.tradeShip.tile()); // Rerender
         break;
       case PathFindResultType.NextTile:
-        // Update safeFromPirates status
+        
+        if (!dstPort.data.pathCache) {
+          dstPort.data.pathCache = new Map<string, TileRef>();
```

Comment:
Can you use currentTile as the key directly?
```suggestion
          dstPort.data.pathCache = new Map<TileRef, TileRef>();
```